File: agentsgym-py/agentsgym_py/markdown.py
from __future__ import annotations

from dataclasses import dataclass

import yaml


@dataclass(frozen=True)
class ParsedMarkdown:
    frontmatter: dict
    content: str


def parse_frontmatter(markdown_text: str) -> ParsedMarkdown:
    """
    Parse YAML frontmatter of the form:

    ---
    key: value
    ---

    ...content...
    """
    text = markdown_text.lstrip("\ufeff")
    lines = text.splitlines()
    if not lines or lines[0].strip() != "---":
        return ParsedMarkdown(frontmatter={}, content=markdown_text)

    end_idx = None
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            end_idx = i
            break

    if end_idx is None:
        return ParsedMarkdown(frontmatter={}, content=markdown_text)

    fm_text = "\n".join(lines[1:end_idx]).strip()
    content = "\n".join(lines[end_idx + 1 :]).lstrip("\n")
    try:
        fm = yaml.safe_load(fm_text) if fm_text else {}
    except Exception:
        fm = {}

    if not isinstance(fm, dict):
        fm = {}

    return ParsedMarkdown(frontmatter=fm, content=content)


def dump_frontmatter(frontmatter: dict) -> str:
    fm_text = yaml.safe_dump(frontmatter, sort_keys=False).strip()
    return f"---\n{fm_text}\n---\n"

File: agentsgym-server/server/utils/markdown.ts
import { parse as parseYaml } from 'yaml'

export type ParsedMarkdown = {
  frontmatter: Record<string, unknown>;
  content: string;
};

export function parseFrontmatter(markdownText: string): ParsedMarkdown {
  const text = markdownText.replace(/^\uFEFF/, '')
  const lines = text.split(/\r?\n/)
  if (lines.length === 0 || lines[0].trim() !== '---') {
    return { frontmatter: {}, content: markdownText }
  }

  let endIdx = -1
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === '---') {
      endIdx = i
      break
    }
  }
  if (endIdx === -1) {
    return { frontmatter: {}, content: markdownText }
  }

  const fmText = lines.slice(1, endIdx).join('\n').trim()
  const content = lines.slice(endIdx + 1).join('\n').replace(/^\n+/, '')
  let fm: unknown = {}
  try {
    fm = fmText ? parseYaml(fmText) : {}
  } catch {
    fm = {}
  }
  if (typeof fm !== "object" || fm === null || Array.isArray(fm)) {
    fm = {}
  }
  return { frontmatter: fm as Record<string, unknown>, content }
}

